{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 4731,
     "status": "ok",
     "timestamp": 1635963377112,
     "user": {
      "displayName": "Olimar Teixeira Borges",
      "photoUrl": "https://lh3.googleusercontent.com/a-/AOh14GgMdjAS3Q63iz7EU5c5xLnicit_3iCeqFP5jZ3__g=s64",
      "userId": "15873534311322573793"
     },
     "user_tz": 180
    },
    "id": "wpUg9ai61ro-",
    "outputId": "1190b445-fa0a-44f0-ea37-da970109494b"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: bioservices in c:\\users\\dellpivotal04\\appdata\\roaming\\python\\python38\\site-packages (1.8.0)\n",
      "Requirement already satisfied: wrapt in c:\\programdata\\anaconda3\\lib\\site-packages (from bioservices) (1.11.2)\n",
      "Requirement already satisfied: beautifulsoup4 in c:\\programdata\\anaconda3\\lib\\site-packages (from bioservices) (4.9.3)\n",
      "Requirement already satisfied: appdirs in c:\\users\\dellpivotal04\\appdata\\roaming\\python\\python38\\site-packages (from bioservices) (1.4.4)\n",
      "Requirement already satisfied: requests-cache in c:\\users\\dellpivotal04\\appdata\\roaming\\python\\python38\\site-packages (from bioservices) (0.8.1)\n",
      "Requirement already satisfied: lxml in c:\\programdata\\anaconda3\\lib\\site-packages (from bioservices) (4.6.1)\n",
      "Requirement already satisfied: pandas in c:\\programdata\\anaconda3\\lib\\site-packages (from bioservices) (1.1.3)\n",
      "Requirement already satisfied: easydev>=0.9.36 in c:\\users\\dellpivotal04\\appdata\\roaming\\python\\python38\\site-packages (from bioservices) (0.12.0)\n",
      "Requirement already satisfied: xmltodict in c:\\programdata\\anaconda3\\lib\\site-packages (from bioservices) (0.12.0)\n",
      "Requirement already satisfied: requests in c:\\users\\dellpivotal04\\appdata\\roaming\\python\\python38\\site-packages (from bioservices) (2.25.1)\n",
      "Requirement already satisfied: colorlog in c:\\users\\dellpivotal04\\appdata\\roaming\\python\\python38\\site-packages (from bioservices) (6.4.1)\n",
      "Requirement already satisfied: suds-jurko in c:\\users\\dellpivotal04\\appdata\\roaming\\python\\python38\\site-packages (from bioservices) (0.6)\n",
      "Requirement already satisfied: grequests in c:\\users\\dellpivotal04\\appdata\\roaming\\python\\python38\\site-packages (from bioservices) (0.6.0)\n",
      "Requirement already satisfied: colorama in c:\\programdata\\anaconda3\\lib\\site-packages (from easydev>=0.9.36->bioservices) (0.4.4)\n",
      "Requirement already satisfied: pexpect in c:\\programdata\\anaconda3\\lib\\site-packages (from easydev>=0.9.36->bioservices) (4.8.0)\n",
      "Requirement already satisfied: soupsieve>1.2 in c:\\programdata\\anaconda3\\lib\\site-packages (from beautifulsoup4->bioservices) (2.0.1)\n",
      "Requirement already satisfied: gevent in c:\\programdata\\anaconda3\\lib\\site-packages (from grequests->bioservices) (20.9.0)\n",
      "Requirement already satisfied: pytz>=2017.2 in c:\\programdata\\anaconda3\\lib\\site-packages (from pandas->bioservices) (2020.1)\n",
      "Requirement already satisfied: python-dateutil>=2.7.3 in c:\\programdata\\anaconda3\\lib\\site-packages (from pandas->bioservices) (2.8.1)\n",
      "Requirement already satisfied: numpy>=1.15.4 in c:\\programdata\\anaconda3\\lib\\site-packages (from pandas->bioservices) (1.19.2)\n",
      "Requirement already satisfied: chardet<5,>=3.0.2 in c:\\programdata\\anaconda3\\lib\\site-packages (from requests->bioservices) (3.0.4)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in c:\\programdata\\anaconda3\\lib\\site-packages (from requests->bioservices) (2020.6.20)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in c:\\users\\dellpivotal04\\appdata\\roaming\\python\\python38\\site-packages (from requests->bioservices) (1.26.5)\n",
      "Requirement already satisfied: idna<3,>=2.5 in c:\\programdata\\anaconda3\\lib\\site-packages (from requests->bioservices) (2.10)\n",
      "Requirement already satisfied: url-normalize<2.0,>=1.4 in c:\\users\\dellpivotal04\\appdata\\roaming\\python\\python38\\site-packages (from requests-cache->bioservices) (1.4.3)\n",
      "Requirement already satisfied: attrs<22.0,>=21.2 in c:\\users\\dellpivotal04\\appdata\\roaming\\python\\python38\\site-packages (from requests-cache->bioservices) (21.2.0)\n",
      "Requirement already satisfied: cattrs<2.0,>=1.8 in c:\\users\\dellpivotal04\\appdata\\roaming\\python\\python38\\site-packages (from requests-cache->bioservices) (1.8.0)\n",
      "Requirement already satisfied: six>=1.5 in c:\\programdata\\anaconda3\\lib\\site-packages (from python-dateutil>=2.7.3->pandas->bioservices) (1.15.0)\n",
      "Requirement already satisfied: zope.interface in c:\\programdata\\anaconda3\\lib\\site-packages (from gevent->grequests->bioservices) (5.1.2)\n",
      "Requirement already satisfied: cffi>=1.12.2 in c:\\programdata\\anaconda3\\lib\\site-packages (from gevent->grequests->bioservices) (1.14.3)\n",
      "Requirement already satisfied: setuptools in c:\\programdata\\anaconda3\\lib\\site-packages (from gevent->grequests->bioservices) (50.3.1.post20201107)\n",
      "Requirement already satisfied: zope.event in c:\\programdata\\anaconda3\\lib\\site-packages (from gevent->grequests->bioservices) (4.5.0)\n",
      "Requirement already satisfied: greenlet>=0.4.17 in c:\\programdata\\anaconda3\\lib\\site-packages (from gevent->grequests->bioservices) (0.4.17)\n",
      "Requirement already satisfied: ptyprocess>=0.5 in c:\\users\\dellpivotal04\\appdata\\roaming\\python\\python38\\site-packages (from pexpect->easydev>=0.9.36->bioservices) (0.7.0)\n",
      "Requirement already satisfied: pycparser in c:\\programdata\\anaconda3\\lib\\site-packages (from cffi>=1.12.2->gevent->grequests->bioservices) (2.20)\n",
      "display.max_colwidth : int or None\n",
      "    The maximum width in characters of a column in the repr of\n",
      "    a pandas data structure. When the column overflows, a \"...\"\n",
      "    placeholder is embedded in the output. A 'None' value means unlimited.\n",
      "    [default: 50] [currently: 400]\n"
     ]
    }
   ],
   "source": [
    "!pip install bioservices\n",
    "from bioservices import UniProt\n",
    "\n",
    "# imports\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import io\n",
    "from scipy import stats\n",
    "import os\n",
    "import json\n",
    "from pandas.io.json import build_table_schema\n",
    "import csv\n",
    "import collections\n",
    "import re\n",
    "import xml.etree.ElementTree as ET\n",
    "import requests\n",
    "\n",
    "pd.set_option('max_rows', 99999)\n",
    "pd.set_option('max_colwidth', 400)\n",
    "pd.describe_option('max_colwidth')\n",
    "\n",
    "\n",
    "class BioDatasets:\n",
    "  def gen_reactome():\n",
    "    reactome = pd.DataFrame()\n",
    "    columns = [\"UNIPROT identifier\", \"Reactome Pathway Stable identifier\", \"URL\", \"Event Name\", \"Evidence code\", \"Species\"]        \n",
    "    for chunk in pd.read_csv('C:\\Datasets\\REACTOME_UniProt2Reactome_All_Levels.txt', header=None, sep = '\\t', chunksize=1000, names=columns):\n",
    "      reactome = pd.concat([reactome, chunk], ignore_index=True)\n",
    "    return reactome \n",
    "\n",
    "  def gen_disgenet():\n",
    "    disgenet = pd.DataFrame()\n",
    "    columns = [\"geneId\",\"geneSymbol\", \"DSI\", \"DPI\", \"diseaseId\", \"diseaseName\", \"diseaseType\", \"diseaseClass\", \n",
    "           \"diseaseSemanticType\", \"score\", \"EI\", \"YearInitial\", \"YearFinal\", \"NofPmids\", \"NofSnps\", \"source\"]\n",
    "    for chunk in pd.read_csv('C:\\Datasets\\DISGENET_curated_gene_disease_associations.tsv', header=0, sep = '\\t', chunksize=1000, comment='#', skiprows=1, names=columns):\n",
    "        disgenet = pd.concat([disgenet, chunk], ignore_index=True)\n",
    "    return pd.DataFrame(disgenet)\n",
    "\n",
    "  def gen_omim():\n",
    "    omim = pd.DataFrame()\n",
    "    columns = [\"Chromosome\", \"Genomic_Position_Start\",\"Genomic_Position_End\",\"Cyto_Location\",\"Computed_Cyto_Location\",\"MIM_Number\",\n",
    "              \"Gene_Symbols\",\"Gene_Name\",\"Approved_Symbol\",\"Entrez_Gene_ID\",\"Ensembl_Gene_ID\",\"Comments\",\"Phenotypes\",\"Mouse_Gene_Symbol/ID\"]\n",
    "    for chunk in pd.read_csv('C:\\Datasets\\OMIM_genemap2.txt', header=0, sep = '\\t', chunksize=1000, comment='#', skiprows=1, names=columns):\n",
    "        omim = pd.concat([omim, chunk], ignore_index=True)\n",
    "    return pd.DataFrame(omim)\n",
    "\n",
    "  def gen_iid_human():\n",
    "    iid_human = pd.DataFrame() \n",
    "    for chunk in pd.read_csv('C:\\Datasets\\IID_human_annotated_PPIs.txt', header=0, sep = '\\t', chunksize=10000):\n",
    "      iid_human = pd.concat([iid_human, chunk], ignore_index=True)\n",
    "    return pd.DataFrame(iid_human)\n",
    "\n",
    "  def gen_drug_bank():\n",
    "    #adapted from source: https://github.com/dhimmel/drugbank\n",
    "    with open('C:\\Datasets\\DRUGBANK_all_full_database.xml', encoding='utf8') as xml_file:\n",
    "        tree = ET.parse(xml_file)\n",
    "    root = tree.getroot()\n",
    "\n",
    "    ns = '{http://www.drugbank.ca}'\n",
    "    inchikey_template = \"{ns}calculated-properties/{ns}property[{ns}kind='InChIKey']/{ns}value\"\n",
    "    inchi_template = \"{ns}calculated-properties/{ns}property[{ns}kind='InChI']/{ns}value\"\n",
    "\n",
    "    rows = list()\n",
    "    for i, drug in enumerate(root):\n",
    "        row = collections.OrderedDict()\n",
    "        assert drug.tag == ns + 'drug'\n",
    "        row['type'] = drug.get('type')\n",
    "        row['drugbank_id'] = drug.findtext(ns + \"drugbank-id[@primary='true']\")\n",
    "        row['name'] = drug.findtext(ns + \"name\")\n",
    "        row['description'] = drug.findtext(ns + \"description\")\n",
    "        row['groups'] = [group.text for group in\n",
    "            drug.findall(\"{ns}groups/{ns}group\".format(ns = ns))]\n",
    "        row['atc_codes'] = [code.get('code') for code in\n",
    "            drug.findall(\"{ns}atc-codes/{ns}atc-code\".format(ns = ns))]\n",
    "        row['categories'] = [x.findtext(ns + 'category') for x in\n",
    "            drug.findall(\"{ns}categories/{ns}category\".format(ns = ns))]\n",
    "        row['inchi'] = drug.findtext(inchi_template.format(ns = ns))\n",
    "        row['inchikey'] = drug.findtext(inchikey_template.format(ns = ns))\n",
    "        \n",
    "        # Add drug aliases\n",
    "        aliases = {\n",
    "            elem.text for elem in \n",
    "            drug.findall(\"{ns}international-brands/{ns}international-brand\".format(ns = ns)) +\n",
    "            drug.findall(\"{ns}synonyms/{ns}synonym[@language='English']\".format(ns = ns)) +\n",
    "            drug.findall(\"{ns}international-brands/{ns}international-brand\".format(ns = ns)) +\n",
    "            drug.findall(\"{ns}products/{ns}product/{ns}name\".format(ns = ns))\n",
    "\n",
    "        }\n",
    "        #aliases.add(row['name'])\n",
    "        row['aliases'] = sorted(aliases)\n",
    "\n",
    "        rows.append(row)\n",
    "    \n",
    "    #DRUGBANK - xml\n",
    "    alias_dict = {row['drugbank_id']: row['aliases'] for row in rows}\n",
    "    with open('C:/Datasets/data/aliases.json', 'w') as fp:\n",
    "        json.dump(alias_dict, fp, indent=2, sort_keys=True)\n",
    "      #DRUGBANK - xml\n",
    "    def collapse_list_values(row):\n",
    "        for key, value in row.items():\n",
    "            if isinstance(value, list):\n",
    "                row[key] = '|'.join(value)\n",
    "        return row\n",
    "\n",
    "    rows = list(map(collapse_list_values, rows))\n",
    "\n",
    "    columns = ['drugbank_id', 'name', 'type', 'groups', 'atc_codes', 'categories', 'inchikey', 'inchi', 'description']\n",
    "    drugbank = pd.DataFrame.from_dict(rows)[columns]\n",
    "    return pd.DataFrame(drugbank)\n",
    "\n",
    "  def gen_drug_bank_protein():\n",
    "    #adapted from source: https://github.com/dhimmel/drugbank\n",
    "    with open('C:\\Datasets\\DRUGBANK_all_full_database.xml', encoding='utf8') as xml_file:\n",
    "        tree = ET.parse(xml_file)\n",
    "    root = tree.getroot()\n",
    "    ns = '{http://www.drugbank.ca}'\n",
    "\n",
    "    #Extract protein information\n",
    "    protein_rows = list()\n",
    "    for i, drug in enumerate(root):\n",
    "        drugbank_id = drug.findtext(ns + \"drugbank-id[@primary='true']\")\n",
    "        for category in ['target', 'enzyme', 'carrier', 'transporter']:\n",
    "            proteins = drug.findall('{ns}{cat}s/{ns}{cat}'.format(ns=ns, cat=category))#.apply(str)\n",
    "            for protein in proteins:\n",
    "                row = {'drugbank_id': drugbank_id, 'category': category}\n",
    "                row['organism'] = protein.findtext('{}organism'.format(ns))\n",
    "                row['known_action'] = protein.findtext('{}known-action'.format(ns))\n",
    "                actions = protein.findall('{ns}actions/{ns}action'.format(ns=ns))\n",
    "                row['actions'] = '|'.join(action.text for action in actions)\n",
    "                uniprot_ids = [polypep.text for polypep in protein.findall(\n",
    "                    \"{ns}polypeptide/{ns}external-identifiers/{ns}external-identifier[{ns}resource='UniProtKB']/{ns}identifier\".format(ns=ns))]            \n",
    "                if len(uniprot_ids) != 1:\n",
    "                    continue\n",
    "                row['uniprot_id'] = uniprot_ids[0]\n",
    "                ref_text = str(protein.findtext(\"{ns}references[@format='textile']\".format(ns=ns)))\n",
    "                pmids = re.findall(r'pubmed/([0-9]+)', ref_text)\n",
    "                row['pubmed_ids'] = '|'.join(pmids)\n",
    "                protein_rows.append(row)\n",
    "\n",
    "    drugbank_protein_df = pd.DataFrame.from_dict(protein_rows)\n",
    "\n",
    "    return pd.DataFrame(drugbank_protein_df)\n",
    "\n",
    "  def gen_mondo():\n",
    "    with open('C:\\Datasets\\MONDO_with_equivalents.json', encoding='utf8') as f:\n",
    "        mondo = json.load(f)\n",
    "    mondo = pd.DataFrame(mondo)    \n",
    "    mondo\n",
    "\n",
    "    #MONDO - json\n",
    "    pd.set_option(\"display.max.columns\", None) # to display all the columns\n",
    "    mondodf = pd.json_normalize(mondo [\"graphs\"], record_path =['nodes'],\n",
    "        meta=['meta', ['definition', 'xrefs', 'synonyms']], errors='ignore')\n",
    "    \n",
    "    return pd.DataFrame(mondodf)\n",
    "  \n",
    "  def gen_drugcentral():\n",
    "    drugcentral = pd.DataFrame()\n",
    "    columns = [\"DRUG_NAME\", \"STRUCT_ID\", \"TARGET_NAME\", \"TARGET_CLASS\", \"ACCESSION\", \"GENE\", \"SWISSPROT\", \"ACT_VALUE\", \n",
    "              \"ACT_UNIT\", \"ACT_TYPE\", \"ACT_COMMENT\", \"ACT_SOURCE\", \"RELATION\", \"MOA\", \"MOA_SOURCE\", \"ACT_SOURCE_URL\", \n",
    "              \"MOA_SOURCE_URL\", \"ACTION_TYPE\", \"TDL\", \"ORGANISM\"]\n",
    "    for chunk in pd.read_csv('C:\\Datasets\\DRUGCENTRAL_drug.target.interaction.tsv', header=0, sep = '\\t', chunksize=1000, names=columns):\n",
    "        drugcentral = pd.concat([drugcentral, chunk], ignore_index=True)\n",
    "    \n",
    "    return pd.DataFrame(drugcentral)\n",
    "\n",
    "  def gen_uniprot():\n",
    "    service = UniProt()  \n",
    "    query = \"reviewed:yes\" #Lists all UniProtKB/Swiss-Prot entries (about UniProtKB).\n",
    "    result = service.search(query, frmt=\"tab\")\n",
    "    uniprotdf = pd.DataFrame()\n",
    "    columns = [\"Entry\", \"Entry name\", \"Status\", \"Protein names\", \"Gene names\", \"Organism\", \"Lenght\"]\n",
    "    for chunk in pd.read_table(io.StringIO(result), header=None, sep = '\\t', chunksize=10000, names=columns):\n",
    "        uniprotdf = pd.concat([uniprotdf, chunk], ignore_index=True)\n",
    "    uniprotdf\n",
    "    \n",
    "    return pd.DataFrame(uniprotdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "accelerator": "TPU",
  "colab": {
   "collapsed_sections": [],
   "name": "BIC_DATASETS.ipynb",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
